network/proxy/mucp/mucp.go

```diff
@@ -79,155 +84,126 @@ func readLoop(r server.Request, s client.Stream) error {
-
-		// recursively execute getRoute
-		return p.getRoute(service)
-	}
-
-	if len(name) == 0 {
-		name = "go.micro.router"
+	// lookup the router
+	results, err := p.Router.Lookup(
+		table.NewQuery(table.QueryService(service)),
+	)
+	if err != nil {
+		return nil, err
 	}
 
-	// lookup the remote router
-
-	var addrs []string
-
-	// set the remote address if specified
-	if len(addr) > 0 {
-		addrs = append(addrs, addr)
-	} else {
-		// we have a name so we need to check the registry
-		services, err := p.Client.Options().Registry.GetService(name)
-		if err != nil {
-			return nil, err
-		}
-
-		for _, service := range services {
-			for _, node := range service.Nodes {
-				addrs = append(addrs, node.Address)
-			}
+	if len(results) > 0 {
+		// update the proxy cache
+		p.Lock()
+		for _, route := range results {
+			p.Routes[service][route.Hash()] = route
 		}
-	}
-
-	// no router addresses available
-	if len(addrs) == 0 {
-		return nil, selector.ErrNoneAvailable
-	}
-
-	var pbRoutes *pb.LookupResponse
-	var gerr error
+		routes = p.Routes[service]
+		p.Unlock()
 
-	// set default client
-	if p.RouterService == nil {
-		p.RouterService = pb.NewRouterService(name, p.Client)
+		return toNodes(routes), nil
 	}
 
-	// TODO: implement backoff and retries
-	for _, addr := range addrs {
-		// call the router
-		proutes, err := p.RouterService.Lookup(context.Background(), &pb.LookupRequest{
-			Query: &pb.Query{
-				Service: service,
-			},
-		}, client.WithAddress(addr))
-		if err != nil {
-			gerr = err
-			continue
-		}
-		// set routes
-		pbRoutes = proutes
-		break
-	}
+	// call the router
+	resp, err := p.RouterService.Lookup(context.Background(), &pb.LookupRequest{
+		Query: &pb.Query{
+			Service: service,
+		},
+	}, p.CallOption)
 
 	// errored out
-	if gerr != nil {
-		return nil, gerr
+	if err != nil {
+		return nil, err
 	}
 
 	// no routes
-	if pbRoutes == nil {
+	if resp == nil || len(resp.Routes) == 0 {
 		return nil, selector.ErrNoneAvailable
 	}
 
 	// convert from pb to []*router.Route
-	for _, r := range pbRoutes.Routes {
-		routes = append(routes, table.Route{
+	for _, r := range resp.Routes {
+		route := table.Route{
 			Service: r.Service,
 			Address: r.Address,
 			Gateway: r.Gateway,
 			Network: r.Network,
 			Link:    r.Link,
 			Metric:  int(r.Metric),
-		})
+		}
+		routes[route.Hash()] = route
 	}
 
 	return toNodes(routes), nil
 }
 
-// ServeRequest honours the server.Router interface
-func (p *Proxy) ServeRequest(ctx context.Context, req server.Request, rsp server.Response) error {
-	// set default client
-	if p.Client == nil {
-		p.Client = client.DefaultClient
+// watchRoutes watches remote router service routes and updates proxy cache
+func (p *Proxy) watchRoutes() {
+	// this is safe to do as the only way watchRoutes returns is
+	// when some error is written into error channel - we want to bail then
+	defer close(p.errChan)
+
+	stream, err := p.RouterService.Watch(context.Background(), &pb.WatchRequest{})
```

Comment:
Should watchRoutes be doing the equivalent of Lookup? Using p.Router.Watch if the interface is present or otherwise p.RouterService.Watch.

Reply:
This is a good question and I must admit it's something I had thought about. The thing is, the `p.Router` as it is **only watches table events when Advertise() has been called**. On its own the router only watches and handles registry events. If you dont `Advertise()` why should you watch **your own routing tables** events? That seemed illogical to me as well as starting `Advertise()` here, hence only the remote watch...